File: adl/devices/reset.py
import logging

from collections import namedtuple

from yapsy.IPlugin import IPlugin

from adl.devices.generic_device import GenericDevice

class Reset(GenericDevice, namedtuple("Reset", ["name"])):

	__slots__ = ()

	@property
	def setup(self):
		return ""

	@property
	def command_handler(self):
		return """
			return {name}.command_handler(command, reply);
		""".format(name=self.cname())

	@property
	def sources(self):
		return ["reset.cpp"]

	@property
	def includes(self):
		return ["reset.h"]

	@property
	def declarations(self):
		return "static Reset {name} = Reset();".format(name=self.cname())

class ResetPlugin(IPlugin):
	def activate(self):
		pass

	def deactivate(self):
		pass

	def get(self, device):
		return Reset(device.name)

	def set_log_level(self, level):
		logging.getLogger(__name__).setLevel(level)

File: adl/devices/reset.cpp
#include <Arduino.h>

#include "device.h"
#include "adl.h"

#include "reset.h"

Reset::Reset() {}

void Reset::reset() {}

void Reset::setup() {}

int Reset::command_handler(char const * const command, char * reply)
{
	int reply_length = 0;
	if (command[0] == 'R')
	{
		int i = 0;
		int n_devices = adl_device_count();

		for (i = 0; i < n_devices; i++)
		{
			adl_get_device(i)->reset();
		}
		
		strcpy(reply, "ROK");
		reply_length = strlen(reply);
	}
	else
	{
		reply[0] = '?';
		reply_length = 1;
	}
	return reply_length;
}
